drivers/gpio/gpio_tca6424a.c

```diff
@@ -0,0 +1,572 @@
+{
+	struct tca6424a_drv_data *drv_data = dev->data;
+	struct tca6424a_irq_state *irq_state = &drv_data->irq_state;
+	int ret;
+	uint32_t previous_state;
+	uint32_t current_state;
+	uint32_t transitioned_pins;
+	uint32_t interrupt_status = 0;
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	/* Any interrupts enabled? */
+	if (!irq_state->rising && !irq_state->falling) {
+		k_sem_give(&drv_data->lock);
+		return;
+	}
+
+	/* Store previous input state then read new value */
+	previous_state = drv_data->pins_state.input;
+	ret = update_input_regs(dev, &current_state);
+	if (ret != 0) {
+		k_sem_give(&drv_data->lock);
+		return;
+	}
+
+	/* Find out which input pins have changed state */
+	transitioned_pins = previous_state ^ current_state;
+
+	/* Mask gpio transactions with rising/falling edge interrupt config */
+	interrupt_status = (irq_state->rising & transitioned_pins & current_state);
+	interrupt_status |= (irq_state->falling & transitioned_pins & previous_state);
+	k_sem_give(&drv_data->lock);
+
+	if (interrupt_status) {
+		gpio_fire_callbacks(&drv_data->callbacks, dev, interrupt_status);
+	}
+}
+
+/**
+ * @brief Work handler for TCA6424A interrupt
+ *
+ * @param work Work struct that contains pointer to interrupt handler function
+ */
+static void tca6424a_work_handler(struct k_work *work)
+{
+	struct tca6424a_drv_data *drv_data = CONTAINER_OF(work, struct tca6424a_drv_data, work);
+
+	tca6424a_handle_interrupt(drv_data->dev);
+}
+
+/**
+ * @brief ISR for interrupt pin of TCA6424A
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param gpio_cb Pointer to callback function struct
+ * @param pins Bitmask of pins that triggered interrupt
+ */
+static void tca6424a_int_gpio_handler(const struct device *dev, struct gpio_callback *gpio_cb,
+				      uint32_t pins)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(pins);
+
+	struct tca6424a_drv_data *drv_data =
+		CONTAINER_OF(gpio_cb, struct tca6424a_drv_data, int_gpio_cb);
+
+	k_work_submit(&drv_data->work);
+}
+
+static int tca6424a_setup_pin(const struct device *dev, gpio_pin_t pin, gpio_flags_t flags)
+{
+	struct tca6424a_drv_data *const drv_data = dev->data;
+	uint32_t reg_cfg = drv_data->pins_state.config;
+	uint32_t reg_out = drv_data->pins_state.output;
+	int ret;
+
+	/* For each pin, 0 == output, 1 == input */
+	if ((flags & GPIO_OUTPUT) != 0) {
+		if ((flags & GPIO_OUTPUT_INIT_HIGH) != 0) {
+			reg_out |= BIT(pin);
+		} else if ((flags & GPIO_OUTPUT_INIT_LOW) != 0) {
+			reg_out &= ~BIT(pin);
+		}
+		ret = update_output_regs(dev, reg_out);
+		if (ret != 0) {
+			return ret;
+		}
+		reg_cfg &= ~BIT(pin);
+	} else {
+		reg_cfg |= BIT(pin);
+	}
+
+	ret = update_config_regs(dev, reg_cfg);
+
+	return ret;
+}
+
+static int tca6424a_pin_config(const struct device *dev, gpio_pin_t pin, gpio_flags_t flags)
+{
+	int ret;
+	struct tca6424a_drv_data *const drv_data = dev->data;
+
+#if (CONFIG_GPIO_LOG_LEVEL >= LOG_LEVEL_DEBUG)
```

Comment:
This should also go if you change the log to dev->name :-)

Reply:
Done